### ISSUE TITLE
Poor inference of dict literal when dict type is unioned with certain types

### ISSUE THREAD

**Issue by henribru** [python/mypy#15217]
**Bug Report**

(A clear and concise description of what the bug is.)

**To Reproduce**

```python
# Ideally, a small sample program that demonstrates the problem.
# Or even better, a reproducible playground link https://mypy-play.net/ (use the "Gist" button)
```
https://mypy-play.net/?mypy=master&python=3.11&gist=5cf1ab002a6dc63c5013fa0f6a00ba4d

**Expected Behavior**
I expect the dict literal to be inferred as `dict[str, str |int]` when passed to both `foo` and `bar`

**Actual Behavior**
It's inferred as `dict[str, object]` when passed to `foo`, which causes a false positive.

Note that the inference depends on the type you union the dict with. E.g. if you replace `Iterable[tuple[str]` with `list[tuple[str]]` you get the expected inference.

**Your Environment**

- Mypy version used: Tried 1.0.0, 1.0.1, 1.1.1, 1.2.0 and master in the playground, all have the same behavior
- Mypy command-line flags: See playground
- Mypy configuration options from `mypy.ini` (and other config files): See playground
- Python version used: 3.11
